### ISSUE TITLE
Kafka should use service DNS names and not Pod IP addresses as advertised host

### ISSUE THREAD

**Issue by scholzj** [EnMasseProject/barnabas#50]
Kafka is currently using the pod IP address as the advertised host name. I think that is not correct, as the Pod IP address is not necessarily stable. It should be using either the IP address exposed through the headless service or probably even better the hostname exposed through the headless service.

I'm not sure in detail how the different Kafka clients work and how often they recheck the advertised addresses. But having them stable should not hurt, it can only help.

**Comment by scholzj** [comment]
It looks like the change to DNS hostnames caused some problems to users who are using Kafka from outside of OpenShift since the hostnames don't resolve outside of OpenShift. We should rollback this change

This is still a valid issue but we should revisit it later when we have other ways how to make Kafka available outside of OpenShift.

**Comment by komljen** [comment]
I agree that pod IP is not a good choice and something like `${HOSTNAME}.kafka-headless` is probably more suitable. Hostname in StatefulSet is stable.

**Comment by scholzj** [comment]
@komljen It was causing some issues when trying to access Kafka from outside of OpenShift. Basically, if you start the OpenShift cluster on Linux using `oc cluster up` you can access the pods directly trhough IP address also from local applications not running on OpenShift. When the advertised hostnames are changed from Pod IP address to DNS names, these DNS names are not resolved outside of OpenShift and this stops working. But some users are using this right now. 

So we need to first implement better way for exposing Kafka to the outside of OpenShift. And once we do that we will again change to using the hostnames instead of the PodIPs.

**Comment by Veryclouder** [comment]
Kafka operator has adopted your advisement, but I can't use svc DNS name to access kafka from outside

**Comment by Veryclouder** [comment]
I hope you can give me some suggestions. 
For example, when I use kafka operator to deploy kafka, I can modify some parameters to use pod ip instead of use DNS name

**Comment by scholzj** [comment]
You should use the external listener to access the Kafka cluster from outside.

**Comment by Veryclouder** [comment]
thanks, from my heart!

**Comment by Veryclouder** [comment]
I'll think about it later. The external listener is used to expose the services inside the cluster. The main problem with Kafka is that the domain name, not the IP, is used to register the backend pod to zookeeper. The domain name cannot be accessed locally. If the domain name resolution is configured locally, it can be accessed, but if the pod IP changes, the configuration file must be changed!



发自 网易邮箱大师




---- 回复的原邮件 ----
| 发件人 | Jakub ***@***.***> |
| 日期 | 2021年11月26日 17:12 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.******@***.***> |
| 主题 | Re: [strimzi/strimzi-kafka-operator] Kafka should use service DNS names and not Pod IP addresses as advertised host (#50) |

You should use the external listener to access the Kafka cluster from outside.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Veryclouder** [comment]
If the external listener is valid, I want to get more messages.How to access kafka by external listener!



发自 网易邮箱大师




---- 回复的原邮件 ----
| 发件人 | Jakub ***@***.***> |
| 日期 | 2021年11月26日 17:12 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.******@***.***> |
| 主题 | Re: [strimzi/strimzi-kafka-operator] Kafka should use service DNS names and not Pod IP addresses as advertised host (#50) |

You should use the external listener to access the Kafka cluster from outside.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
